examples/consume-body-in-extractor-or-middleware/src/main.rs

```diff
@@ -0,0 +1,97 @@
+//! Run with
+//!
+//! ```not_rust
+//! cargo run -p example-consume-body-in-extractor-or-middleware
+//! ```
+
+use axum::{
+    async_trait,
+    body::{self, BoxBody, Bytes, Full},
+    extract::{FromRequest, RequestParts},
+    http::{Request, StatusCode},
+    middleware::{self, Next},
+    response::{IntoResponse, Response},
+    routing::post,
+    Router,
+};
+use std::net::SocketAddr;
+use tower::ServiceBuilder;
+use tower_http::ServiceBuilderExt;
+use tracing_subscriber::{layer::SubscriberExt, util::SubscriberInitExt};
+
+#[tokio::main]
+async fn main() {
+    tracing_subscriber::registry()
+        .with(tracing_subscriber::EnvFilter::new(
+            std::env::var("RUST_LOG")
+                .unwrap_or_else(|_| "example_consume_body_in_extractor_or_middleware=debug".into()),
+        ))
+        .with(tracing_subscriber::fmt::layer())
+        .init();
+
+    let app = Router::new().route("/", post(handler)).layer(
+        ServiceBuilder::new()
+            .map_request_body(body::boxed)
+            .layer(middleware::from_fn(print_request_body)),
+    );
+
+    let addr = SocketAddr::from(([127, 0, 0, 1], 3000));
+    tracing::debug!("listening on {}", addr);
+    axum::Server::bind(&addr)
+        .serve(app.into_make_service())
+        .await
+        .unwrap();
+}
+
+// middleware that shows how to consume the request body upfront
+async fn print_request_body(
+    request: Request<BoxBody>,
+    next: Next<BoxBody>,
+) -> Result<impl IntoResponse, Response> {
+    let request = buffer_request_body(request).await?;
+
+    Ok(next.run(request).await)
+}
+
+// extractor that shows how to consume the request body upfront
+struct PrintRequestBody;
+
+#[async_trait]
+impl FromRequest<BoxBody> for PrintRequestBody {
+    type Rejection = Response;
+
+    async fn from_request(req: &mut RequestParts<BoxBody>) -> Result<Self, Self::Rejection> {
+        let request = Request::from_request(req)
+            .await
+            .map_err(|err| err.into_response())?;
+
+        let request = buffer_request_body(request).await?;
+
+        *req = RequestParts::new(request);
```

Comment:
This business with recreating the `RequestParts` is necessary because `RequestParts::body_mut` returns `Option<&mut B>` and not `&mut Option<B>`. That means once you've taken the body, you cannot put it back. I'm thinking we should change that for 0.5.

Reply:
Seems reasonable.

Reply:
I'll do that in a separate PR then 😊